packages/outdated/src/outdated.ts

```diff
@@ -0,0 +1,153 @@
+
+export interface OutdatedPackage {
+  alias: string,
+  belongsTo: DependenciesField,
+  current?: string, // not defined means the package is not installed
+  latestManifest?: PackageManifest,
+  packageName: string,
+  wanted: string,
+}
+
+export default async function outdated (
+  opts: {
+    compatible?: boolean,
+    currentLockfile: Lockfile | null,
+    getLatestManifest: GetLatestManifestFunction,
+    include?: IncludedDependencies,
+    lockfileDir: string,
+    manifest: ProjectManifest,
+    match?: (dependencyName: string) => boolean,
+    prefix: string,
+    wantedLockfile: Lockfile,
+  },
+): Promise<OutdatedPackage[]> {
+  if (packageHasNoDeps(opts.manifest)) return []
+  const allDeps = getAllDependenciesFromPackage(opts.manifest)
+  const importerId = getLockfileImporterId(opts.lockfileDir, opts.prefix)
+  const currentLockfile = opts.currentLockfile || { importers: { [importerId]: {} } }
+
+  const outdated: OutdatedPackage[] = []
+
+  await Promise.all(
+    DEPENDENCIES_FIELDS.map(async (depType) => {
+      if (
+        opts.include?.[depType] === false ||
+        !opts.wantedLockfile.importers[importerId][depType]
+      ) return
+
+      let pkgs = Object.keys(opts.wantedLockfile.importers[importerId][depType]!)
+
+      if (opts.match) {
+        pkgs = pkgs.filter((pkgName) => opts.match!(pkgName))
+      }
+
+      await Promise.all(
+        pkgs.map(async (alias) => {
+          const ref = opts.wantedLockfile.importers[importerId][depType]![alias]
+
+          // ignoring linked packages. (For backward compatibility)
+          if (ref.startsWith('file:')) {
+            return
+          }
+
+          const relativeDepPath = dp.refToRelative(ref, alias)
+
+          // ignoring linked packages
+          if (relativeDepPath === null) return
+
+          const pkgSnapshot = opts.wantedLockfile.packages?.[relativeDepPath]
+
+          if (!pkgSnapshot) {
+            throw new Error(`Invalid ${WANTED_LOCKFILE} file. ${relativeDepPath} not found in packages field`)
+          }
+
+          const currentRef = currentLockfile.importers[importerId] &&
+            currentLockfile.importers[importerId][depType] &&
+            currentLockfile.importers[importerId][depType]![alias]
+          const currentRelative = currentRef && dp.refToRelative(currentRef, alias)
+          const current = currentRelative && dp.parse(currentRelative).version || currentRef
+          const wanted = dp.parse(relativeDepPath).version || ref
+          const { name: packageName } = nameVerFromPkgSnapshot(relativeDepPath, pkgSnapshot)
+
+          // It might be not the best solution to check for pkgSnapshot.name
+          // TODO: add some other field to distinct packages not from the registry
+          if (pkgSnapshot.resolution && (pkgSnapshot.resolution['type'] || pkgSnapshot.name)) { // tslint:disable-line:no-string-literal
```

Comment:
I'm curious why you are using `pkgSnapshot.resolution['type']` and not `pkgSnapshot.resolution.type`.

Reply:
Probably it was failing that way. This type is a union of several types. One of the types doesn't have the `type` field.

Reply:
So if you reference a non-existent field in a union, typescript fails instead of returning undefined?

Reply:
yes. There's a workaround to add `type?: undefined` to the type that doesn't have this field. But I think that's worse